argo/src/Argo/ServerState.hs

```diff
@@ -0,0 +1,180 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+module Argo.ServerState (ServerState, StateID, getAppState, initialState, freshStateReader, appStateCache, stateRecipe, stateRecipes, initialStateID, newStateID, recipeFile) where
+
+import Control.Lens
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import Data.Hashable (Hashable)
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Aeson as JSON
+import Data.Text (Text)
+import Data.IORef
+import Data.UUID (UUID)
+import qualified Data.UUID as UUID
+import qualified Data.UUID.V4 as UUID
+
+import GHC.IO.Exception
+
+import qualified Crypto.Hash.SHA1 as SHA1 (hash)
+
+newtype StateID = StateID (Maybe UUID) deriving (Eq, Ord, Show, Hashable)
+
+initialStateID :: StateID
+initialStateID = StateID Nothing
+
+newStateID :: IO StateID
+newStateID = StateID . Just <$> UUID.nextRandom
+
+instance JSON.ToJSON StateID where
+  toJSON (StateID (Just uuid)) = JSON.String (UUID.toText uuid)
+  toJSON (StateID Nothing) = JSON.Null
+
+instance JSON.FromJSON StateID where
+  parseJSON JSON.Null = pure $ StateID Nothing
+  -- Note: this is a backwards compatibility hack that should be removed soon
```

Comment:
Should we emit a warning if this is used? Should we just break compatibility now, given that we currently are the only users?

Reply:
We are not the only users - I know of at least one other. I think we should break the backwards compat soon, though.